src/neural/encoder.cc

```diff
@@ -36,6 +36,26 @@ const int kPlanesPerBoard = 13;
 const int kAuxPlaneBase = kPlanesPerBoard * kMoveHistory;
 }  // namespace
 
+void SetInputPlanesFromBoard(InputPlanes& planes, int board_index,
+                             const ChessBoard& board, int repetitions) {
+  const int base = board_index * kPlanesPerBoard;
+  planes[base + 0].mask = (board.ours() * board.pawns()).as_int();
+  planes[base + 1].mask = (board.our_knights()).as_int();
+  planes[base + 2].mask = (board.ours() * board.bishops()).as_int();
+  planes[base + 3].mask = (board.ours() * board.rooks()).as_int();
+  planes[base + 4].mask = (board.ours() * board.queens()).as_int();
+  planes[base + 5].mask = (board.our_king()).as_int();
+
+  planes[base + 6].mask = (board.theirs() * board.pawns()).as_int();
+  planes[base + 7].mask = (board.their_knights()).as_int();
+  planes[base + 8].mask = (board.theirs() * board.bishops()).as_int();
+  planes[base + 9].mask = (board.theirs() * board.rooks()).as_int();
+  planes[base + 10].mask = (board.theirs() * board.queens()).as_int();
+  planes[base + 11].mask = (board.their_king()).as_int();
+
+  if (repetitions >= 1) planes[base + 12].SetAll();
+}
+
 InputPlanes EncodePositionForNN(const PositionHistory& history,
```

Comment:
Instead of making a copy of a board and modifying it, do something like:
(untested)
```cpp
InputPlanes EncodePositionForNN(const PositionHistory& history,
                                int history_planes) {
  InputPlanes result(kAuxPlaneBase + 8);

  {
    const ChessBoard& board = history.Last().GetBoard();
    const bool we_are_black = board.flipped();
    if (board.castlings().we_can_000()) result[kAuxPlaneBase + 0].SetAll();
    if (board.castlings().we_can_00()) result[kAuxPlaneBase + 1].SetAll();
    if (board.castlings().they_can_000()) result[kAuxPlaneBase + 2].SetAll();
    if (board.castlings().they_can_00()) result[kAuxPlaneBase + 3].SetAll();
    if (we_are_black) result[kAuxPlaneBase + 4].SetAll();
    result[kAuxPlaneBase + 5].Fill(history.Last().GetNoCaptureNoPawnPly());
    // Plane kAuxPlaneBase + 6 used to be movecount plane, now it's all zeros.
    // Plane kAuxPlaneBase + 7 is all ones to help NN find board edges.
    result[kAuxPlaneBase + 7].SetAll();
  }

  bool flip = false;
  int history_idx = history.GetLength() - 1;
  for (int i = 0; i < std::min(history_planes, kMoveHistory);
       ++i, --history_idx) {
    if (history_idx < 0 && !fill_empty_history) break;
    const Position& position =
        history.GetPositionAt(history_idx < 0 ? 0 : history_idx);
    const ChessBoard& board =
        flip ? position.GetThemBoard() : position.GetBoard();

    const int base = i * kPlanesPerBoard;
    result[base + 0].mask = (board.ours() * board.pawns()).as_int();
    result[base + 1].mask = (board.our_knights()).as_int();
    result[base + 2].mask = (board.ours() * board.bishops()).as_int();
    result[base + 3].mask = (board.ours() * board.rooks()).as_int();
    result[base + 4].mask = (board.ours() * board.queens()).as_int();
    result[base + 5].mask = (board.our_king()).as_int();

    result[base + 6].mask = (board.theirs() * board.pawns()).as_int();
    result[base + 7].mask = (board.their_knights()).as_int();
    result[base + 8].mask = (board.theirs() * board.bishops()).as_int();
    result[base + 9].mask = (board.theirs() * board.rooks()).as_int();
    result[base + 10].mask = (board.theirs() * board.queens()).as_int();
    result[base + 11].mask = (board.their_king()).as_int();

    const int repetitions = position.GetRepetitions();
    if (repetitions >= 1) result[base + 12].SetAll();

    if (history_idx < 0 && board.en_passant()) {
      const idx = GetLowestBit(board.en_passant);
      if (idx < 8) {  // "Us" board
        result[base + 0].mask +
            ((0x0000000000000100ULL - 0x0000000001000000ULL) << idx);
      } else {
        result[base + 6].mask +
            ((0x0001000000000000ULL - 0x0000000100000000ULL) << (idx - 56));
      }
    }
    if (history_idx > 0) flip = !flip;
  }

  return result;
}
```

as diff:
```diff
diff --git a/src/neural/encoder.cc b/src/neural/encoder.cc
index d25f6c8..0784bcf 100644
--- a/src/neural/encoder.cc
+++ b/src/neural/encoder.cc
@@ -57,9 +57,10 @@ InputPlanes EncodePositionForNN(const PositionHistory& history,
   bool flip = false;
   int history_idx = history.GetLength() - 1;
   for (int i = 0; i < std::min(history_planes, kMoveHistory);
-       ++i, flip = !flip, --history_idx) {
-    if (history_idx < 0) break;
-    const Position& position = history.GetPositionAt(history_idx);
+       ++i, --history_idx) {
+    if (history_idx < 0 && !fill_empty_history) break;
+    const Position& position =
+        history.GetPositionAt(history_idx < 0 ? 0 : history_idx);
     const ChessBoard& board =
         flip ? position.GetThemBoard() : position.GetBoard();
 
@@ -80,6 +81,18 @@ InputPlanes EncodePositionForNN(const PositionHistory& history,
 
     const int repetitions = position.GetRepetitions();
     if (repetitions >= 1) result[base + 12].SetAll();
+
+    if (history_idx < 0 && board.en_passant()) {
+      const idx = GetLowestBit(board.en_passant);
+      if (idx < 8) {  // "Us" board
+        result[base + 0].mask +
+            ((0x0000000000000100ULL - 0x0000000001000000ULL) << idx);
+      } else {  // "Them" board
+        result[base + 6].mask +
+            ((0x0001000000000000ULL - 0x0000000100000000ULL) << (idx - 56));
+      }
+    }
+    if (history_idx > 0) flip = !flip;
   }
 
   return result;
```